app/ide-desktop/lib/dashboard/mock/authentication/src/authentication/cognito.ts

```diff
@@ -0,0 +1,775 @@
+     * password, new password, and repeat new password to change their old password to the new
+     * one. The validation of the repeated new password is handled by the `changePasswordModel`
+     * component. */
+    changePassword(oldPassword: string, newPassword: string) {
+        return changePassword(oldPassword, newPassword)
+    }
+
+    /** We want to signal to Amplify to fire a "custom state change" event when the user is
+     * redirected back to the application after signing in via an external identity provider. This
+     * is done so we get a chance to fix the location history. The location history is the history
+     * of the pages visited within the application. Amplify messes up the history when it redirects
+     * the user to the identity provider's authentication page. This is because Amplify believes
+     * that we are in the browser, so the location needs to be modified to account for leaving the
+     * page and coming back. However, in the Electron app we never leave the page. The rest of the
+     * flow is handled in the system browser instead. So we must undo the changes that Amplify
+     * makes.
+     *
+     * In order to do so, we need to pass custom state along for the entire OAuth flow, which is
+     * obtained by calling this function. This function will return the current location path if
+     * the user is signing in from the desktop application, and `null` otherwise.
+     *
+     * We use `null` outside of the desktop application because Amplify only messes up the
+     * location history in the desktop application.
+     *
+     * See: https://github.com/aws-amplify/amplify-js/issues/3391#issuecomment-756473970 */
+    private customState() {
+        return detect.isOnElectron() ? window.location.pathname : null
+    }
+}
+
+// ===================
+// === UserSession ===
+// ===================
+
+/** User's session, provides information for identifying and authenticating the user. */
+export interface UserSession {
+    /** User's email address, used to uniquely identify the user.
+     *
+     * Provided by the identity provider the user used to log in. One of:
+     *
+     * - GitHub,
+     * - Google, or
+     * - Email. */
+    email: string
+    /** User's access token, used to authenticate the user (e.g., when making API calls). */
+    accessToken: string
+}
+
+/** Parse a {@link cognito.CognitoUserSession} into a {@link UserSession}.
+ * @throws If the `email` field of the payload is not a string. */
+function parseUserSession(session: cognito.CognitoUserSession): UserSession {
+    const payload: Record<string, unknown> = session.getIdToken().payload
+    const email = payload.email
+    /** The `email` field is mandatory, so we assert that it exists and is a string. */
+    if (typeof email !== 'string') {
+        throw new Error('Payload does not have an email field.')
+    } else {
+        const accessToken = `aJwtToken.${btoa(JSON.stringify({ username: email }))}.`
```

Comment:
What's this `aJwtToken.` bit? A mock JWT header? Should probably put it in a constant with some docs.

Reply:
for now i've just removed the header instead, doesn't seem like it's needed

Reply:
for now i've just removed the header instead, doesn't seem like it's needed